### ISSUE TITLE
Network Health not running

### ISSUE THREAD

**Issue by distnz** [jonfairbanks/OctoPrint-NetworkHealth#15]
I have installed via plugin manager, and run the configuration command however the network health plugin does not appear to function. I have checkd my logs and find the following:

2021-05-08 07:23:08,260 - octoprint.plugins.networkhealth - ERROR - Could not run network health check
Traceback (most recent call last):
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_NetworkHealth/__init__.py", line 18, in _check_network
    if not self.check_ping():
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_NetworkHealth/__init__.py", line 54, in check_ping
    response = os.system("ping -c 4 " + hostname)
TypeError: can only concatenate str (not "NoneType") to str

and it just keeps looping.
Any suggestions?

**Comment by jonfairbanks** [comment]
It looks like the plugin is failing to get your default gateway IP (your router).

https://github.com/jonfairbanks/OctoPrint-NetworkHealth/blob/master/octoprint_NetworkHealth/__init__.py#L53

- What device/OS?
- Does your network have a unique setup?

**Comment by distnz** [comment]
Version 0.17.0, running on Raspberry Pi 4 Model B Rev 1.1
Nothing unique about my network, just 3 Asus routers running their stock OS, running AIMesh. the onyl wifi issue i have is this damn Pi and it just falling off the network randomly. Was hoping this scirpt would help like the others it has helped :P

**Comment by jonfairbanks** [comment]
Can you please create a `test.py` file with these contents:
```
import socket
import struct

def default_gateway():
    with open("/proc/net/route") as fh:
        for line in fh:
            fields = line.strip().split()
            if fields[1] != '00000000' or not int(fields[3], 16) & 2:
                continue
            return socket.inet_ntoa(struct.pack("<L", int(fields[2], 16)))

print("Checking default gateway...")
gateway = default_gateway()
print(gateway)
```

Then run with `python test.py` and you should hopefully see your router IP address. If you see something different please share a screenshot of the script output.

**Comment by distnz** [comment]
![test py](https://user-images.githubusercontent.com/3527272/117519801-530cd700-aff9-11eb-9703-603b84eece4d.png)
It dosent return anything unfortunatley

**Comment by jonfairbanks** [comment]
We'll just get it straight from Linux. Please share the output of `ip route | grep default`.

**Comment by distnz** [comment]
Here we go

default via 192.168.1.1 dev wlan0 proto dhcp src 192.168.1.120 metric 303

**Comment by jonfairbanks** [comment]
Can you please uninstall and add a test version which should handle this scenario.

- Remove previous versions of NetworkHealth
- Go to Plugin Manager > Get More
- Install from URL: `https://github.com/jonfairbanks/OctoPrint-NetworkHealth/archive/develop.zip`
- After reboot; confirm in the logs the plugin is running.

For your use-case I have added a catch that if the default gateway cannot be determined it'll fallback to pinging Google. You may see the below message in octoprint.log:
```
Failed to determine default gateway; using 8.8.8.8 instead...
```

If this works for you please report back and I will push a fixed release to the plugins repo.

**Comment by distnz** [comment]
Excellent, Just done this now, Confirmed in logs v1.0.4 and that it has
started doing its thing "INFO - Beginning to monitor Network health..."

Will report back in a few hours. Ill keep an eye on the logs to see if i
can see it saving my drops etc :)

Thanks for your fast responses and assistance, it is greatly appreciated :)

**Comment by distnz** [comment]
So just went to check and the printer is pffline, ive rebooted it and pulled the logs:

2021-05-08 16:56:40,294 - octoprint.util.comm - INFO - Changing monitoring state from "Finishing" to "Operational"
2021-05-08 16:56:59,826 - octoprint.server.heartbeat - INFO - Server heartbeat <3
2021-05-08 17:11:59,826 - octoprint.server.heartbeat - INFO - Server heartbeat <3
2021-05-08 17:26:59,827 - octoprint.server.heartbeat - INFO - Server heartbeat <3
2021-05-08 17:41:59,828 - octoprint.server.heartbeat - INFO - Server heartbeat <3
2021-05-08 17:56:59,829 - octoprint.server.heartbeat - INFO - Server heartbeat <3
2021-05-08 18:11:59,830 - octoprint.server.heartbeat - INFO - Server heartbeat <3
2021-05-08 18:20:54,292 - octoprint.plugins.networkhealth - ERROR - No Network Connection - Resetting Adapter(s)...
2021-05-08 18:24:51,441 - octoprint.plugins.networkhealth - INFO - Failed to determine default gateway; using 8.8.8.8 instead...
2021-05-08 18:24:51,480 - octoprint.plugins.networkhealth - ERROR - No Network Connection - Resetting Adapter(s)...
2021-05-08 18:26:59,831 - octoprint.server.heartbeat - INFO - Server heartbeat <3
2021-05-08 18:27:37,256 - octoprint.util.connectivity.connectivity_checker - INFO - Connectivity changed from online to offline
2021-05-08 18:27:37,256 - octoprint.util.connectivity.connectivity_checker - INFO - Connecting to 1.1.1.1:53 is not working
2021-05-08 18:27:37,257 - octoprint.util.connectivity.connectivity_checker - INFO - Resolving octoprint.org is not working
2021-05-08 18:28:15,973 - octoprint.plugins.networkhealth - INFO - Failed to determine default gateway; using 8.8.8.8 instead...
2021-05-08 18:28:16,021 - octoprint.plugins.networkhealth - ERROR - No Network Connection - Resetting Adapter(s)...
2021-05-08 18:31:41,786 - octoprint.plugins.networkhealth - INFO - Failed to determine default gateway; using 8.8.8.8 instead...
2021-05-08 18:31:41,825 - octoprint.plugins.networkhealth - ERROR - No Network Connection - Resetting Adapter(s)...
2021-05-08 18:34:59,347 - octoprint.server.util.sockjs - INFO - Client connection closed: ::ffff:192.168.1.20
2021-05-08 18:34:59,951 - octoprint.plugins.networkhealth - INFO - Failed to determine default gateway; using 8.8.8.8 instead...
2021-05-08 18:34:59,979 - octoprint.plugins.networkhealth - ERROR - No Network Connection - Resetting Adapter(s)...
2021-05-08 18:35:23,070 - octoprint.server.util.sockjs - INFO - Client connection closed: ::ffff:192.168.1.236
2021-05-08 18:38:23,135 - octoprint.plugins.networkhealth - INFO - Failed to determine default gateway; using 8.8.8.8 instead...
2021-05-08 18:38:24,450 - octoprint.plugins.networkhealth - ERROR - No Network Connection - Resetting Adapter(s)...
2021-05-08 18:41:52,847 - octoprint.plugins.networkhealth - INFO - Failed to determine default gateway; using 8.8.8.8 instead...
2021-05-08 18:41:54,451 - octoprint.plugins.networkhealth - ERROR - No Network Connection - Resetting Adapter(s)...
2021-05-08 18:42:12,454 - octoprint.server.heartbeat - INFO - Server heartbeat <3
2021-05-08 18:45:20,274 - octoprint.plugins.networkhealth - INFO - Failed to determine default gateway; using 8.8.8.8 instead...
2021-05-08 18:45:20,308 - octoprint.plugins.networkhealth - ERROR - No Network Connection - Resetting Adapter(s)...
2021-05-08 18:48:40,976 - octoprint.plugins.networkhealth - INFO - Failed to determine default gateway; using 8.8.8.8 instead...
2021-05-08 18:48:41,330 - octoprint.plugins.networkhealth - ERROR - No Network Connection - Resetting Adapter(s)...

So it looks like the plugin is working, but it never actually manages ot reset the adaptors, somethings stopping it coming back up.

**Comment by jonfairbanks** [comment]
Based on the timestamps it does appear to be recovering for a short amount of time. 

These log entries are interesting though:
```
2021-05-08 18:34:59,347 - octoprint.server.util.sockjs - INFO - Client connection closed: ::ffff:192.168.1.20
```
```
2021-05-08 18:35:23,070 - octoprint.server.util.sockjs - INFO - Client connection closed: ::ffff:192.168.1.236
```

What devices are 192.168.1.20 & 192.168.1.236?

To simulate what the plugin would do during network loss you could run `sudo ip link set wlan0 down && sudo ip link set wlan0 up`. This should *not* require you to provide a sudo password and should give no output if successful once ran.

**Comment by distnz** [comment]
Ah thos are my Garage computer and y Laptop (i would have been connecting
to the pi to test the various things, grab logs etc) and didnt close when
done :P

Ill give that simulation test a run shortly, once the current print has
finished :)
